File: MAQ Software/46.permutations.go
package leetcode

// Permute46 chooses one unused number for each position. A bool slice gives O(1)
// used checks; the path is copied only at complete permutations.
//
// Time: O(n!*n). Space: O(n) excluding output.
func Permute46(nums []int) [][]int {
	ans := [][]int{}
	path := []int{}
	used := make([]bool, len(nums))
	var dfs func()
	dfs = func() {
		if len(path) == len(nums) {
			ans = append(ans, append([]int(nil), path...))
			return
		}
		for i, v := range nums {
			if used[i] {
				continue
			}
			used[i] = true
			path = append(path, v)
			dfs()
			path = path[:len(path)-1]
			used[i] = false
		}
	}
	dfs()
	return ans
}

File: companies/Oracle/46.permutations.cpp
#include <algorithm>
#include <cctype>
#include <climits>
#include <cmath>
#include <cstdlib>
#include <functional>
#include <numeric>
#include <queue>
#include <sstream>
#include <string>
#include <unordered_map>
#include <utility>
#include <vector>
using namespace std;


class Solution {
public:
    vector<vector<int>> permute(vector<int>& nums) {
        /*
        Approach:
        Backtrack by choosing each unused value for the next position. A used
        array marks which indices are already in the current permutation.

        Complexity: O(n! * n) time including output copies, O(n) stack space.
        */
        vector<vector<int>> ans;
        vector<int> path;
        vector<bool> used(nums.size(), false);
        function<void()> dfs = [&]() {
            if (path.size() == nums.size()) {
                ans.push_back(path);
                return;
            }
            for (int i = 0; i < (int)nums.size(); ++i) {
                if (used[i]) continue;
                used[i] = true;
                path.push_back(nums[i]);
                dfs();
                path.pop_back();
                used[i] = false;
            }
        };
        dfs();
        return ans;
    }
};
